plan/physical_initialize.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+)
+
+type physicalInitializer struct {
+	ctx       context.Context
+	allocator *idAllocator
+}
+
+func (ps *physicalInitializer) initialize(p PhysicalPlan) {
+	for _, child := range p.GetChildren() {
+		ps.initialize(child.(PhysicalPlan))
+
+		// Initialize the attribute that will be affected by its children.
+		p.SetCorrelated(p.IsCorrelated() || child.IsCorrelated())
+	}
+	switch pp := p.(type) {
+	case *PhysicalAggregation:
```

Comment:
I mean the Logical Opt Phase has set the agg's correlation. If you need to set it here, remove the set correlated logic in aggregation push down and selection push down. And then consider all kinds of operators to set correlation. You can abstract it in interface.